Universal classifier with data-reuploading

quantshah opened the pull request:
**Description of the Change:**

Added an example which implements a simple Neural Network with a data-uploading circuit discussed in [https://arxiv.org/abs/1907.02085](https://arxiv.org/abs/1907.02085)

I am a bit confused about how the plots are rendered?
@josh146

josh146 commented:
This is great, thanks @quantshah! Have you run `make docs` to compile the documentation? Sphinx gallery will run the Python script, and automatically embed the plots created under each `plt.show()` command

quantshah commented:
Oh I see. Running it now and updating the PR with the latest plots.

quantshah commented:
Hi,

I think this is ready for review. If you want to see how it looks then check this pdf that I generated locally:

[Universal Quantum Classifier — PennyLane 0.4.0 documentation.pdf](https://github.com/XanaduAI/pennylane/files/3432018/Universal.Quantum.Classifier.PennyLane.0.4.0.documentation.pdf)

quantshah commented:
Something cool would be to show how the decision boundary keeps updating with training as an animation/gif. But maybe I will add it later.

josh146 commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,491 @@
+In order to load data onto a single qubit, we use a unitary operation U
+(:math:`x_1`, :math:`x_2`, :math:`x_3`) which is just a parameterized
+matrix multiplication representing the rotation of the state vector in
+the Bloch sphere. Eg., to load :math:`(x_1, x_2)` into the qubit, we
```
```suggestion
the Bloch sphere. E.g., to load :math:`(x_1, x_2)` into the qubit, we
```

josh146 commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,491 @@
+(:math:`x_1`, :math:`x_2`, :math:`x_3`) which is just a parameterized
+matrix multiplication representing the rotation of the state vector in
+the Bloch sphere. Eg., to load :math:`(x_1, x_2)` into the qubit, we
+just start from some initial state vector, :math:`|0 \rangle `,
```
No spaces are allowed before the backtick in math mode, otherwise it will cause parts of the paragraph not to render
```suggestion
just start from some initial state vector, :math:`|0 \rangle`,
```

josh146 commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,491 @@
+just start from some initial state vector, :math:`|0 \rangle `,
+apply the unitary operation :math:`U(x_1, x_2, 0)` and end up at a new
+point on the Bloch sphere. Here we have padded 0 since our data is only
+2D. Adri´an P´erez-Salinas et al. (2019) discuss how to load a higher
```
Need to fix the accented letters (as above)

josh146 commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,491 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Once we load the data onto the quantum circuit, we want to have some
+trainable nonlinear model similar to a Neural Network and a way of
```
```suggestion
trainable nonlinear model similar to a neural network as well as a way of
```

josh146 commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,491 @@
+Once we load the data onto the quantum circuit, we want to have some
+trainable nonlinear model similar to a Neural Network and a way of
+learning the weights of the model from data. This is again done with
+unitaries, :math:`U(\theta_1, \theta_2, \theta_3)` such that we load the
```
```suggestion
unitaries, :math:`U(\theta_1, \theta_2, \theta_3)`, such that we load the
```

josh146 commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,491 @@
+is much like having a set of ouput neurons and selecting the one which
+has the highest value as the label.
+
+In ``pennylane``, we can define an observable (the expected output
```
```suggestion
In PennyLane, we can define an observable (the expected output
```

josh146 commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,491 @@
+some intuitive idea why, read “Why does deep and cheap learning work so
+well?” by Henry W. Lin, Max Tegmark (MIT) and David Rolnick (2016).
+
+Adri´an P´erez-Salinas et al. (2019) describe that in their approach the
```
accents

josh146 commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,491 @@
+
+[1] Pérez-Salinas, Adrián, et al. “Data re-uploading for a universal
+quantum classifier.” arXiv preprint arXiv:1907.02085 (2019).
+
```
Might make more sense in terms of the flow to place the references at the end of the tutorial

josh146 commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,491 @@
+
+.. figure:: ../../examples/figures/universal_layers.png
+   :scale: 75%
+   :alt: Layers
```
The style of these images is great!

For reference:

![image](https://user-images.githubusercontent.com/2959003/61889958-07c72880-af39-11e9-9df1-7a4bfe70f8d4.png)

josh146 commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,491 @@
+[1] Pérez-Salinas, Adrián, et al. “Data re-uploading for a universal
+quantum classifier.” arXiv preprint arXiv:1907.02085 (2019).
+
+“Talk is cheap. Show me the code. - Linus Torvalds”
```
```suggestion
"Talk is cheap. Show me the code." - Linus Torvalds
```

josh146 commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,491 @@
+
+##############################################################################
+# Make a dataset of points in and out of a circle
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
These feel like they should be sub headings (or just plain text/sentences describing each code-block) rather than top-level page headings

josh146 commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,491 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+"""
+
+import pennylane as pl
```
For consistency, best to use `qml` throughout the code, rather than `pl`

josh146 commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,491 @@
+        
+    radius: float
+        The radius of the circle.
+    """
```
For consistency with the later parts of the tutorial, might be good to standardize all docstrings to the same style (my preference is Google style as below, e.g., `Args:`, rather than NumPy style, as it reduces the line count and tends to be easier to parse/read)

josh146 commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,491 @@
+
+##############################################################################
+# `pennylane.about()`
+# ~~~~~~~
```
Maybe a bit superfluous as a heading, could just be replaced with a sentence along the lines of "This tutorial was executed with the following PennyLane version and environment"

quantshah commented on doc/tutorials.rst:
```diff
@@ -66,6 +66,11 @@ interesting QML problems in PennyLane!
+    :tooltip: Universal Quantum Classifier with data-reuploading
+    :figure: ../examples/figures/universal_dnn.png
+    :description: :ref:`universal_classifier`
+
```
Thanks. Perfect size!

quantshah commented:
Thanks for the very thorough review @josh146. I took care of all the comments. I would just mention that using `pl` as the abbreviation for pennylane makes more sense to me than `qml`. Since we also have a qml toolbox. Would it be possible to change *all* the qml references (in docs, code, examples etc.) to `pl` sometime? 

@mariaschuld If you could please have a check on the explanation part. I read the paper and tried to explain it to the best of my understanding.

mariaschuld commented:
Hey @quantshah, will most likely only get to it after Wednesday, but I spoke to one of the authors of the paper recently and it is great to have this implemented so quickly!

josh146 commented:
>  I would just mention that using pl as the abbreviation for pennylane makes more sense to me than qml.

Definitely something worth discussing. Using `qml` is a naming convention used before PennyLane acquired its existing name. Until we make a decision to change the naming convention across the board, however, best to keep everything consistent.

mariaschuld commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,474 @@
+Transforming quantum states using unitary operations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A single-qubit quantum state is characterized by a two-dimensional state
```
This is nice for beginners, but a bit strange if a user has read through a lot of tutorials already that do much more advanced things with qubits, only to hear now what a qubit is...I'd consider removing, or rephrasing as "As a reminder..." and shorten to one sentence.

mariaschuld commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,474 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In order to load data onto a single qubit, we use a unitary operation
+:math:`U(x_1, x_2, x_3)` which is just a parameterized
```
Do you want to say what `U` exactly is, i.e. a qml.Rot(...) operation?

mariaschuld commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,474 @@
+one or other class. This happens when we measure the quantum state which
+leads to its projection as either the state 0 or 1. We can compute the
+fidelity (or closeness) of the output state to the class label making
+the output state jump to either :math:`| 0 \rangle` or
```
Again, this is a wonderful beginner's explanation, but VERY basic...

mariaschuld commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,474 @@
+accuracy. But in practice, it might require a large number of neurons in
+the single hidden layer and here is where Deep Neural Networks come into
+action. Deep Neural Networks proved to be better in practice and we have
+some intuitive idea why, read “Why does deep and cheap learning work so
```
I'd be careful, there are lots of explanations, especially from the last two years that are not mentioned in that paper...

mariaschuld commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,474 @@
+
+label_0 = [[1.0 + 0.0j], [0.0 + 0.0j]]
+label_1 = [[0.0 + 0.0j], [1.0 + 0.0j]]
+state_labels = [label_0, label_1]
```
Do you not want to code it so that labels are real, as in normal ML?

mariaschuld commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,474 @@
+r"""
+.. _universal_classifier:
+
+Universal Quantum Classifier 
```
Can we change the name to 'Single-qubit quantum classifier'? "universal" can confuse people, since lots of qclassifiers are computationally universal, but not as function approximators...

co9olguy replied:
Agreed. What about something "Data-reuploading classifer" to reflect the paper content?

quantshah replied:
Done

mariaschuld commented on examples/pennylane_universal_classifier.py:
```diff
@@ -0,0 +1,474 @@
+        float: fidelity between output state and input
+    """
+    for v in var:
+        qml.Rot(*x, wires=0)
```
In the paper the version for which universality is proven feeds each feature in as w*x+b, or not?

quantshah replied:
Yes. I changed it now but in principle the alternating application of data unitaries and weight unitaries should work. I am also not trying to exactly follow their paper and reproduce their code but more taking the idea/inspiration. For example, they define the cost as sum of fidelities. I have another version where I take those fidelities and define some kind of a cross entropy loss (multiple class). Thanks for the comment though.

mariaschuld commented:
Also, could we turn this into a template, together with the most important options mentioned in the paper? It should be almost trivial, but requires careful test-writing and documentation...

co9olguy commented:
Hi all, any updates on the status of this PR? Looks like there were some requested changes/comments that need to be addressed?

josh146 commented:
@quantshah said he will be working on this this week.

Once #291 is merged, we will need to change the target of this PR to master.

co9olguy commented:
In regards to @mariaschuld's comments, I do agree that this is a more basic introduction, but I think that's ok. 

We can certainly make sure that the "core" tutorials build off one another. 

But it also makes sense to have under the "QML algorithms" section of the gallery different tutorials that are not necessarily sequential or dependent on one another. For instance, we could mark some tutorials as "Beginner" or "Introductory" and others as "Advanced" or "Expert". A user might end up at this tutorial as the first one they clicked on, depending on what catches their eye.

josh146 commented on examples/pennylane_skip_plugins_hybrid.py:
```diff
@@ -0,0 +1,391 @@
+r"""
```
Ah. I merged master into this branch, but it didn't rename the old `skip` tutorials. They'll have to be deleted manually, I suppose?

quantshah commented:
Hi everyone, thank you so much for the careful and detailed review. I addressed some of the comments and rebased against latest master. I am doing much more with this code in a personal project but this should be good enough for anyone to start playing with the idea presented in the original paper.

This is ready for a re-review. Do I need to manually remove the "skip" tutorials? I am a bit confused about that.

Also, in the original approach, the authors train using L-BFGS (which I do not mention here anymore as @co9olguy suggested it might get confusing), but I use ADAM. Now the issue is, I am using only a small number of epochs and with each run, the decision boundary is slightly different even though the accuracy is really good. This might also be due to using a very simple fidelity based cost function instead of the "weighted fidelity" that the authors use.

Sorry for the delay here since I myself am trying to understand this work better at the theory level.

josh146 commented:
> This is ready for a re-review. Do I need to manually remove the "skip" tutorials? I am a bit confused about that.

Previously, we used `skip` in filenames to make a tutorial as not to be executed by Sphinx gallery. Now, we instead use an explicit `run` in the filename to mark those that _should_ be run. Those without `run` are implicitly skipped.

However it looks like a previous merge caused the old files to reappear in this PR; could you delete the following:
```
examples/pennylane_QGAN.py 
examples/pennylane_advanced_usage.py 
examples/pennylane_gaussian_transformation.py
examples/pennylane_qubit_rotation.py
examples/pennylane_skip_plugins_hybrid.py 
examples/pennylane_skip_pytorch_noise.py
examples/pennylane_skip_quantum_neural_net.py 
examples/pennylane_skip_state_preparation.py
examples/pennylane_skip_variational_classifier.py
examples/pennylane_variational_quantum_eigensolver.py 
```

In addition, could you revert the following files to the version in master:
```console
git checkout origin/master doc/_static/thumbs/code.png
git checkout origin/master pennylane/decorator.py
```

Thanks!

co9olguy commented on examples/pennylane_run_universal_classifier.py:
```diff
@@ -317,7 +319,8 @@ def test(weights, x, y, state_labels=None):
     dm_labels = [density_matrix(s) for s in state_labels]
     for i in range(len(x)):
-        expectation = qcircuit(weights, x=x[i], y=dm_labels[y[i]])
+        zvals = params[0] + params[1]*x[i]
```
Just doing a fly-by review at the moment, but this caught my eye. What led to this reparameterization?

quantshah replied:
This is the one that the paper uses. The basic idea is to introduce data and weights in the circuit using unitaries. So, instead of having a data unitary and then a weight unitary, they combine it in this way and apply it together. @mariaschuld mentioned this in a previous review. I am not sure if I want to completely following their implementation word by word though. Eg., they use a weighted fidelity function as los and here I am not using that.

quantshah commented:
Hi @co9olguy and @mariaschuld I am waiting for your approval on this. Sometimes the training goes really well and we get upto 91% in the test data accuracy (even more than what they got in the paper)
Epoch: 30 | Loss: 0.111283 | Train accuracy: 0.880000 | Test accuracy: 0.917500.

![Figure_1](https://user-images.githubusercontent.com/6968324/64622194-3749cb80-d3e7-11e9-8477-4d8ee6dcd967.png)

This is of course dependent on the hyper parameters / regularisation etc. So at this point, I am not so worried about reproducing the exact results of the paper. But rather just showing this new way of loading data into a Quantum Neural Network (via the unitaries).

Works decently with other types of data too:

![plot](https://user-images.githubusercontent.com/6968324/64622425-9dcee980-d3e7-11e9-835b-6b1d5b3293c6.png)

co9olguy commented on examples/pennylane_run_data_reuploading_classifier.py:
```diff
@@ -0,0 +1,482 @@
+the output label using the `Hermitian <https://pennylane.readthedocs.io/en/latest/code/ops/qubit.html#pennylane.ops.qubit.Hermitian>`_
+operator. The expectation value of the observable gives the overlap or fidelity.
+We can then define the cost function as the sum of the fidelities for all
+the data points after passing throuh the circuit and optimize the parameters
```
```suggestion
the data points after passing through the circuit and optimize the parameters
```

co9olguy commented on examples/pennylane_run_data_reuploading_classifier.py:
```diff
@@ -0,0 +1,482 @@
+optimal weights for classification. Gradient based optimizers such as
+Adam (Kingma et. al., 2014) can be used if we have a good model of
+the circuit and how noise might affect it. Or, we can use some
+gradient-free method such as L-BFGS (Liu, Dong C., and Jorge Nocedal, 1989)
```
```suggestion
gradient-free method such as L-BFGS (Liu, Dong C., and Nocedal, J., 1989)
```

co9olguy commented on examples/pennylane_run_data_reuploading_classifier.py:
```diff
@@ -0,0 +1,482 @@
+
+Now, we can use our favorite optimizer to maximize the sum of the
+fidelities over all data points (or batches of datapoints) and find the
+optimal weights for classification. Gradient based optimizers such as
```
```suggestion
optimal weights for classification. Gradient-based optimizers such as
```

co9olguy commented on examples/pennylane_run_data_reuploading_classifier.py:
```diff
@@ -0,0 +1,482 @@
+# [1] Pérez-Salinas, Adrián, et al. “Data re-uploading for a universal
+# quantum classifier.” arXiv preprint arXiv:1907.02085 (2019).
+
+# [2] Kingma, Diederik P., and Jimmy Ba. "Adam: A method for stochastic
```
```suggestion
# [2] Kingma, Diederik P., and Ba, J. "Adam: A method for stochastic
```

co9olguy commented on examples/pennylane_run_data_reuploading_classifier.py:
```diff
@@ -0,0 +1,482 @@
+# [2] Kingma, Diederik P., and Jimmy Ba. "Adam: A method for stochastic
+# optimization." arXiv preprint arXiv:1412.6980 (2014).
+
+# [3] Liu, Dong C., and Jorge Nocedal. "On the limited memory BFGS
```
```suggestion
# [3] Liu, Dong C., and Nocedal, J. "On the limited memory BFGS
```

co9olguy commented on examples/pennylane_run_data_reuploading_classifier.py:
```diff
@@ -0,0 +1,482 @@
+
+# [3] Liu, Dong C., and Jorge Nocedal. "On the limited memory BFGS
+# method for large scale optimization." Mathematical programming 
+# 45.1-3 (1989): 503-528.
```
```suggestion
# 45.1-3 (1989): 503-528.

```
